### DISCUSSION TITLE
Missing environment variable leads to "process.env.VARIABLE" in the bundle

### DISCUSSION THREAD

**Discussion by tmcdos** [quasarframework/quasar#16680] [General - Components / Directives / etc]
If I assign a value to an environment variable, let's say `VARIABLE=test` then anywhere in the code where there is `process.env.VARIABLE` it gets replaced with the value of that environment variable.
However, if the environment variable is not defined - Webpack or Quasar or whatever is responsible for the replacement of `process.env.VARIABLE` with its actual value does not replace it and then `process.env.VARIABLE` ends up in the final bundle and obviously throws an error in the browser because there is no `process` variable.

For example, this code
```javascript
  const createHistory = process.env.SERVER || process.env.WEB_COMPONENT
    ? createMemoryHistory
    : (process.env.VUE_ROUTER_MODE === 'history' ? createWebHistory : createWebHashHistory);
```
is compiled by Webpack to this
```javascript
const createHistory =  false ? 0 :  true ? vue_router__WEBPACK_IMPORTED_MODULE_3__.createWebHistory : 0;
```
if the environment variables `SERVER` and `WEB_COMPONENT` are defined.

However, if `WEB_COMPONENT` is missing/undefined - the compiled code is
```javascript
const createHistory =  false || process.env.WEB_COMPONENT ? vue_router__WEBPACK_IMPORTED_MODULE_3__.createMemoryHistory :  true ? vue_router__WEBPACK_IMPORTED_MODULE_3__.createWebHistory : 0;
```
I would expect that in this case the `process.env.WEB_COMPONENT` is replaced with `undefined` (or at least an empty string) instead of left as is:
```javascript
const createHistory =  false || undefined ? vue_router__WEBPACK_IMPORTED_MODULE_3__.createMemoryHistory :  true ? vue_router__WEBPACK_IMPORTED_MODULE_3__.createWebHistory : 0;
```

What can I do to achieve the behavior I expect instead of the current behavior?
One workaround would be to define all possible/used environment variables with some default value in `quasar.config.js` like
```javascript
env:
{
  SERVER: process.env.SERVER || false,
  WEB_COMPONENT: process.env.WEB_COMPONENT || false,
}
```
But it still feels as a workaround rather than a true solution.
Is this the only way ?
If it can be solved by patching `@quasar/app` - I can patch it locally.

**Comment by tmcdos** [comment]
Looks like `dotenv-webpack` solves this problem by stubbing the missing variables - i.e. if the environment `VARIABLE` is missing then the `process.env` part of `process.env.VARIABLE` will be replaced with `"MISSING_ENV_VAR"` because
```javascript
"MISSING_ENV_VAR".VARIABLE
```
is a valid JavaScript code.
It this possible somehow with `@quasar/app` ?

**Comment by tmcdos** [comment]
In the end my solution goes like this:
- not adding anything to the `env` object inside `quasar.config.js`
- adding `dotenv-webpack` as a plugin inside `extendWebpack` in `quasar.config.js`

```javascript
const EnvPlugin = require('./webpack-env-plugin');
module.exports = function(ctx)
{
  return {
    env: {},
    extendWebpack(cfg)
    {
       cfg.plugins.push(new EnvPlugin({ systemvars: true }));
    }
  };
};
```

**Comment by rstoenescu** [comment]
You can always define all env vars and assign boolean false to those that are "disabled".

**Reply by tmcdos** [reply]
Always defining everything is an artificial requirement and serves as a workaround for the issue. Compiling a working code even in the absence of some environment variable(s) is the correct solution.
